### ISSUE TITLE
Incompatible with vite

### ISSUE THREAD

**Issue by mikecousins** [react-csv/react-csv#268]
I've been playing around with migrating from create react app to vite and all of my packages seem to be working except react-csv.

```
npm run dev

> freddie-web@1.0.0 dev
> vite

Pre-bundling dependencies:
  react
  react-dom
  @sentry/react
  @sentry/tracing
  axios
  (...and 32 more)
(this will be run only when your dependencies or config have changed)
 > node_modules/react-csv/src/components/Link.js:110:6: error: Unexpected "<"
    110 │       <a
        ╵       ^

error when starting dev server:
Error: Build failed with 1 error:
node_modules/react-csv/src/components/Link.js:110:6: error: Unexpected "<"
    at failureErrorWithLog (C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:1198:15)
    at buildResponseToResult (C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:923:32)
    at C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:1022:20
    at C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:569:9
    at handleIncomingPacket (C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:658:9)
    at Socket.readFromStdout (C:\GitHub\freddie\front-end\node_modules\esbuild\lib\main.js:536:7)
    at Socket.emit (node:events:378:20)
    at addChunk (node:internal/streams/readable:313:12)
```

**Comment by todd-elvers** [comment]
Just ran into this issue as well.  I believe this stems from how Vite treats JS files that contain React but have the suffix `.js`.  As proof, if I navigate to `node_modules/react-csv/src/components` and change `Link.js` to `Link.jsx` Vite will make it past the error.

Some research turned up [this Github issue in Vite](https://github.com/vitejs/vite/issues/769) where the author indicates that processing `.jsx` files is more expensive than `.js` files, which is why Vite makes the assumption that `.js` => JS loader and `.jsx` => JSX loader.  

Sadly that issue is closed and the author seems unwilling to provide a way to specify that some JS files should be processed by the JSX loader.  What's even more bizarre is that the underlying library `esbuild` supports a CLI flag that can force `.js` files to be treated like `.jsx` files, but Vite does not allow access to that CLI argument.

I guess for now maybe a `postinstall` step or something that renames that file?  Feels super dirty though when renaming `Link.js` to `Link.jsx` fixes the issue completely.

**Comment by mikecousins** [comment]
Yeah sounds like vite isn't going to move on this one. That's a pretty harsh comment at the end.

**Comment by matzeso** [comment]
Just came here to add a PR to change the filename. Although I am unhappy with the way Evan / Vite dealt with this issue, I think overall he is right in that it makes no sense to put JSX in JS files and just by that trash the incredible performance boost that you can get from native modules.

There was a long discussion in the AirBnB Styleguide on Github (its linked in the aforementioned issue) and the author over there also said (+1 for his ability to look ahead) that they would not put JSX within .JS files because they are different syntax and not Standard compliant and looking at native modules that would result in problems. 

I think it would be wise to look to the future here, as native module resolution probably becomes the standard in the foreseeable future.
Or am I missing any downsides in renaming the link to `.jsx`?

**Comment by akshay-developer-2021** [comment]
My old codebase have a lot of JSX in in JS files. Now If I need to use Vite I have to rename all files ?

**Comment by mscolnick** [comment]
You can also add this esbuild plugin: 

```typescript
import path from 'path';
import type { Plugin } from 'esbuild';

export function fixReactCsvPlugin(): Plugin {
  return {
    name: 'resolve-fixup',
    setup: (build) => {
      build.onResolve({ filter: /react-csv$/ }, () => {
        return Promise.resolve({
          path: path.join(process.cwd(), './node_modules/react-csv/index.js'),
        });
      });
    },
  };
}
```

**Comment by urswtlvin** [comment]
I have same problems...

I add @mscolnick 's plugin to my project's `vite.config.ts` but it is not working...
Browser's console outputs this error: (I'm using 5000port for vite)
```
GET http://localhost:5000/@fs/node_modules/react-csv/index.js net::ERR_ABORTED 404 (Not Found)
```

``` vite.config.ts
import { defineConfig } from "vite";
import reactRefresh from "@vitejs/plugin-react-refresh";
import fixReactCsvPlugin from "./fixReactCsvPlugin";

export default defineConfig({
    root: "./",
    plugins: [reactRefresh(), fixReactCsvPlugin()],
    server: {
        port: 5000,
        open: true,
    },
});
```

Do you have any ideas to fix this problem?

**Comment by Njima1572** [comment]
> Just ran into this issue as well. I believe this stems from how Vite treats JS files that contain React but have the suffix `.js`. As proof, if I navigate to `node_modules/react-csv/src/components` and change `Link.js` to `Link.jsx` Vite will make it past the error.
> 
> Some research turned up [this Github issue in Vite](https://github.com/vitejs/vite/issues/769) where the author indicates that processing `.jsx` files is more expensive than `.js` files, which is why Vite makes the assumption that `.js` => JS loader and `.jsx` => JSX loader.
> 
> Sadly that issue is closed and the author seems unwilling to provide a way to specify that some JS files should be processed by the JSX loader. What's even more bizarre is that the underlying library `esbuild` supports a CLI flag that can force `.js` files to be treated like `.jsx` files, but Vite does not allow access to that CLI argument.
> 
> I guess for now maybe a `postinstall` step or something that renames that file? Feels super dirty though when renaming `Link.js` to `Link.jsx` fixes the issue completely.

I made a for for this fix, so if anyone is desparate of using it with vite, you can do so by
`yarn add njima1572/react-csv#03ff6edf1921c275f406a8ae76f15f94b0a3856d`
or 
`npm i njima1572/react-csv#03ff6edf1921c275f406a8ae76f15f94b0a3856d`

I know this is not a legit fix, but will be a work around until it gets fixed.

**Comment by chindathorn** [comment]
@Njima1572 
@a2441918 
I have the same problem

**Comment by pbeshai** [comment]
Seems there's an issue having both the optimizeDeps and resolve config when using dev, so the plugin produces that @fs 404 error that @urswtlvin mentioned. This seems to fix it (note the change to defineConfig to use a conditional config to get `command` "serve" vs "build"– could alternatively use `mode` but this works).

```js
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import path from 'path';

// https://vitejs.dev/config/
export default defineConfig(({ command }) => {
  return { plugins: [react(), fixReactCsvPlugin(command)], server: {} };
});

function fixReactCsvPlugin(command) {
  const esbuildPlugin = {
    name: 'esbuild:resolve-fixes',
    setup: (build) => {
      build.onResolve({ filter: /react-csv$/ }, (args) => {
        return Promise.resolve({
          path: path.join(process.cwd(), './node_modules/react-csv/index.js'),
        });
      });
    },
  };

  return {
    name: 'resolve-fixes',
    config() {
      // only use the esbuild plugin during dev ("serve" command)
      if (command === 'serve') {
        return {
          optimizeDeps: {
            // development fixes
            esbuildOptions: {
              plugins: [esbuildPlugin],
            },
          },
        };
      }

      // build
      return {
        resolve: {
          // production fixes
          alias: [
            {
              find: 'react-csv',
              replacement: './node_modules/react-csv/index.js',
            },
          ],
        },
      };
    },
  };
}
```

**Comment by abdennour** [comment]
I merged #269 [but CI failed](https://github.com/react-csv/react-csv/runs/4577026326?check_suite_focus=true). I reverted back.
If you have a green fix, please open another PR
